src/types_.py

```diff
@@ -3,17 +3,332 @@
 
 """Types for uploading docs to charmhub."""
 
-
+import dataclasses
 import typing
+from enum import Enum
+from pathlib import Path
+
+Content = str
+Url = str
 
 
 class Page(typing.NamedTuple):
     """Information about a documentation page.
 
-    Atrs:
+    Attrs:
         url: The link to the page.
         content: The documentation text of the page.
     """
 
-    url: str
-    content: str
+    url: Url
+    content: Content
+
+
+NavlinkTitle = str
+
+
+class IndexFile(typing.NamedTuple):
+    """Information about a documentation page.
+
+    Attrs:
+        title: The title for the index.
+        content: The local content of the index.
+    """
+
+    title: NavlinkTitle
+    content: Content | None
+
+
+class Index(typing.NamedTuple):
+    """Information about the local and server index page.
+
+    Attrs:
+        server: The index page on the server.
+        local: The local index file contents.
+    """
+
+    server: Page | None
+    local: IndexFile
+
+
+Level = int
+TablePath = str
+
+
+class PathInfo(typing.NamedTuple):
+    """Represents a file or directory in the docs directory.
+
+    Attrs:
+        local_path: The path to the file on the local disk.
+        level: The number of parent directories to the docs folder including the docs folder.
+        table_path: The computed table path based on the disk path relative to the docs folder.
+        navlink_title: The title of the navlink.
+    """
+
+    local_path: Path
+    level: Level
+    table_path: TablePath
+    navlink_title: NavlinkTitle
+
+
+PathInfoLookup = dict[tuple[Level, TablePath], PathInfo]
+
+
+class Navlink(typing.NamedTuple):
+    """Represents navlink of a table row of the navigation table.
+
+    Attrs:
+        title: The title of the documentation page.
+        link: The relative URL to the documentation page or None if there is no link.
+    """
+
+    title: NavlinkTitle
+    link: str | None
+
+
+class TableRow(typing.NamedTuple):
+    """Represents one parsed row of the navigation table.
+
+    Attrs:
+        level: The number of parents, is 1 if there is no parent.
+        path: The a unique string identifying the row.
+        navlink: The title and relative URL to the documentation page.
+    """
+
+    level: Level
+    path: TablePath
+    navlink: Navlink
+
+    @property
+    def is_group(self) -> bool:
+        """Whether the row is a group of pages."""
+        return self.navlink.link is None
+
+    def to_line(self) -> str:
```

Comment:
What do you think about that it seems to clash with the purpose of the `__str__` in Python? https://docs.python.org/3/reference/datamodel.html#object.__str__

Specifically, "to compute the “informal” or nicely printable string representation of an object". The `to_line` method is intended to be the representation of a `TableRow` specifically on the navigation table rather than be a representation of `TableRow`. It would change, for example, what is printed and logged from:

```Python
print(TableRow(1, "path 1", ("title 1", "link 1")))
TableRow(level=1, path='path 1', navlink=('title 1', 'link 1'))
```

To:

```Python
print(TableRow(1, "path 1", Navlink("title 1", "link 1")))
| 1 | path 1 | [title 1](link 1) |
```

It might be less helpful in the case of debugging

Reply:
Ok, it does make sense, perhaps then renaming the method to `to_markdown` as the name of the class already states that it's a `TableRow`

Reply:
Wouldn't we use __repr__() for debugging ?

Reply:
That is a much better method name, thanks!

`__repr__` also has a specific purpose which is to represent a code snippet that would result in the object being generated. It is a useful debugging tool as well, also quite valuable for logs

Reply:
Marking as resolved for now, please let me know if there is any further feedback